### PR TITLE
Simplify node discovery providers

### PR DESCRIPTION
**Pull request by kuujo**
This PR refactors node discovery providers to move failure detection solely under the purview of `ClusterMembershipService` per #868

### CODE REVIEW

File: cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterMembershipService.java [hunk 1]
```diff
@@ -152,118 +157,138 @@ private void handleDiscoveryEvent(NodeDiscoveryEvent event) {
...
+      members.put(remoteMember.id(), remoteMember);
+      post(new ClusterMembershipEvent(ClusterMembershipEvent.Type.MEMBER_ADDED, remoteMember));
+    } else {
+      // If the member's metadata changed, overwrite the local member. We do this instead of a version check since
+      // a heartbeat from the source member always takes precedence over a more recent version.
+      if (!Objects.equals(member.version(), remoteMember.version())
+          || !Objects.equals(member.zone(), remoteMember.zone())
+          || !Objects.equals(member.rack(), remoteMember.rack())
+          || !Objects.equals(member.host(), remoteMember.host())
+          || !Objects.equals(member.properties(), remoteMember.properties())) {
```
Comment (johnou) [inline cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterMembershipService.java]: forget which type properties is, should we be using Arrays.equals here or Objects.equals does the job?

File: utils/src/main/java/io/atomix/utils/config/ConfigMapper.java [hunk 2]
```diff
@@ -558,20 +573,24 @@ protected static String toSetterName(String name) {
   protected static class SetterDescriptor {
     private final String name;
     private final Method setter;
+    private final boolean deprecated;
 
     SetterDescriptor(String name, Method setter) {
       this.name = name;
       this.setter = setter;
+      this.deprecated = setter.getAnnotation(Deprecated.class) != null;
```
Comment (johnou) [inline utils/src/main/java/io/atomix/utils/config/ConfigMapper.java]: nice!

Comment (kuujo) [reply utils/src/main/java/io/atomix/utils/config/ConfigMapper.java]: I know, I really like that feature. We need to log a lot better information about configuration options. Another nice feature would be making suggestions for invalid properties, e.g. using Levenshtein distance to match a real property or something. `Did you mean 'cluster.members.1.id'?`

File: cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterMembershipService.java (hunk 1 above)
Comment (kuujo) [reply cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterMembershipService.java]: It's a `Properties` object

**Pull request merged**